CRM/Mailing/BAO/Mailing.php

```diff
@@ -260,21 +262,38 @@ public static function getRecipients($mailingID) {
         ->groupBy(array("$contact.id", "$email.id"))
         ->replaceInto($includedTempTablename, array('contact_id', 'email_id'))
         ->param('#groups', $recipientsGroup['Include'])
+        ->param('#mailingID', $mailingID)
         ->execute();
     }
 
     if (count($includeSmartGroupIDs)) {
-      CRM_Utils_SQL_Select::from($contact)
-        ->select("$contact.id as contact_id, $email.id as email_id")
-        ->join($email, " INNER JOIN $email ON $email.contact_id = $contact.id ")
-        ->join('gc', " INNER JOIN civicrm_group_contact_cache gc ON gc.contact_id = $contact.id ")
-        ->join('temp', " LEFT JOIN $excludeTempTablename temp ON temp.contact_id = $contact.id ")
-        ->where('gc.group_id IN (#groups)')
-        ->where($includeFilters)
-        ->orderBy($order_by)
-        ->replaceInto($includedTempTablename, array('contact_id', 'email_id'))
-        ->param('#groups', $includeSmartGroupIDs)
-        ->execute();
+      if ($isSMSmode) {
+        CRM_Utils_SQL_Select::from($contact)
+          ->select("$contact.id as contact_id, $phone.id as phone_id")
+          ->join($phone, " INNER JOIN $phone ON $phone.contact_id = $contact.id ")
+          ->join('gc', " INNER JOIN civicrm_group_contact_cache gc ON $contact.id = gc.contact_id ")
+          ->join('temp', " LEFT JOIN $excludeTempTablename temp ON $contact.id = temp.contact_id ")
+          ->where('gc.group_id IN (#groups)')
+          ->where($includeFilters)
+          ->replaceInto($includedTempTablename, array('contact_id', 'phone_id'))
+          ->param('#groups', $includeSmartGroupIDs)
+          ->param('#mailingID', $mailingID)
+          ->execute();
+      }
+      else {
+        CRM_Utils_SQL_Select::from($contact)
+          ->select("$contact.id as contact_id, $email.id as email_id")
+          ->join($email, " INNER JOIN $email ON $email.contact_id = $contact.id ")
+          ->join('gc', " INNER JOIN civicrm_group_contact_cache gc ON gc.contact_id = $contact.id ")
+          ->join('temp', " LEFT JOIN $excludeTempTablename temp ON temp.contact_id = $contact.id ")
+          ->where('gc.group_id IN (#groups)')
+          ->where($includeFilters)
+          ->orderBy($order_by)
+          ->replaceInto($includedTempTablename, array('contact_id', 'email_id'))
+          ->param('#groups', $includeSmartGroupIDs)
+          ->param('#mailingID', $mailingID)
+          ->execute();
+      }
```

Comment:
This is almost all duplicated - could we just store the label of the ID field in a variable to use? This should help avoid SMS being accidentally omitted if anything else were to change. Or do you think that would make it unnecessarily hard to read?